internal/controller/operator/factory/vmagent/collect_scrapes_test.go

```diff
@@ -52,402 +45,346 @@ func TestSelectServiceMonitors(t *testing.T) {
 			},
-			predefinedObjects: predefinedObjects,
-			want:              []string{"other2/ss2"},
-			wantErr:           false,
 		},
-	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			fclient := k8stools.GetTestClientWithObjects(tt.predefinedObjects)
-			got, err := selectServiceScrapes(context.TODO(), tt.args.p, fclient)
-			if (err != nil) != tt.wantErr {
-				t.Errorf("SelectServiceScrapes() error = %v, wantErr %v", err, tt.wantErr)
-				return
-			}
-			gotNames := []string{}
-			for _, monitorName := range got {
-				gotNames = append(gotNames, fmt.Sprintf("%s/%s", monitorName.Namespace, monitorName.Name))
-			}
-			sort.Strings(gotNames)
-			if !reflect.DeepEqual(gotNames, tt.want) {
-				t.Errorf("SelectServiceScrapes() got = %v, want %v", gotNames, tt.want)
-			}
-		})
-	}
+		predefinedObjects: predefinedObjects,
+		want:              []string{"other1/ss1"},
+	})
+
+	// if NamespaceSelector and Selector both defined, selectAllByDefault=true and empty watchNS
+	f(opts{
+		cr: &vmv1beta1.VMAgent{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "default-agent",
+				Namespace: "default",
+			},
+			Spec: vmv1beta1.VMAgentSpec{
+				ServiceScrapeNamespaceSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"name": "other2"}},
+				ServiceScrapeSelector:          &metav1.LabelSelector{MatchLabels: map[string]string{"name": "ss2"}},
+				SelectAllByDefault:             true,
+			},
+		},
+		predefinedObjects: predefinedObjects,
+		want:              []string{"other2/ss2"},
+	})
 }
 
 func TestSelectPodMonitors(t *testing.T) {
-	type args struct {
-		p *vmv1beta1.VMAgent
-		l logr.Logger
-	}
-	tests := []struct {
-		name              string
-		args              args
+	type opts struct {
+		cr                *vmv1beta1.VMAgent
 		want              []string
-		wantErr           bool
 		predefinedObjects []runtime.Object
-	}{
-		{
-			name: "selector pod scrape at vmagent ns",
-			args: args{
-				p: &vmv1beta1.VMAgent{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "example-vmagent",
-						Namespace: "default",
-					},
-					Spec: vmv1beta1.VMAgentSpec{
-						PodScrapeSelector: &metav1.LabelSelector{},
-					},
-				},
-				l: logf.Log.WithName("unit-test"),
-			},
-			predefinedObjects: []runtime.Object{
-				&vmv1beta1.VMPodScrape{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "pod1",
-						Namespace: "default",
-					},
-					Spec: vmv1beta1.VMPodScrapeSpec{},
+	}
+	f := func(o opts) {
+		fclient := k8stools.GetTestClientWithObjects(o.predefinedObjects)
+		got, err := selectPodScrapes(context.TODO(), o.cr, fclient)
+		if err != nil {
+			t.Errorf("SelectPodScrapes() error = %v", err)
+			return
+		}
+		var gotNames []string
+
+		for _, k := range got {
+			gotNames = append(gotNames, fmt.Sprintf("%s/%s", k.Namespace, k.Name))
+		}
+		sort.Strings(gotNames)
+		if !reflect.DeepEqual(gotNames, o.want) {
+			t.Errorf("SelectPodScrapes() got = %v, want %v", gotNames, o.want)
```

Comment:
```suggestion
			t.Errorf("SelectServiceScrapes(): %s", cmp.Diff(gotNames, o.want))
```
Diffs are more informative